### PR TITLE
Config nested

### PR DESCRIPTION
**Pull request by freddiejbawden**
~ Copy of #20, should bypass CodeCov checks ~
## Description
Enable modules configurations.
 [*Click up link*](https://app.clickup.com/t/2enget)

## Changes
- Configuration submodules declarations are auto-generated
- configuration files support `#include` functionality
- configuration files are split into submodules

### CODE REVIEW

**Comment by freddiejbawden** [comment]
Okay, once reviewed activate admin powers and merge through the codecov blocks.

File: Makefile
Comment (brano2) [inline Makefile:114]: What does this do? Is `ar` a bash command?

In general, there is a fair bit of bash in the Makefile which not too many people are familiar with. Do you think it would be a good idea to add some comments to explain how stuff works? If so, maybe add one for this line.

File: configurations/config.txt
Comment (brano2) [inline configurations/config.txt:3]: BOOOOO 👎 

(jk, no need to change it now)

File: configurations/config.txt
Comment (brano2) [inline configurations/config.txt:7]: Where are the module names defined? Maybe add that to the comment.

File: test/src/utils/config.test.cpp
Comment (brano2) [inline test/src/utils/config.test.cpp:68]: You could also test some failure cases like circular includes or typos in names of files/modules/fields.

File: utils/config.mk
Comment (brano2) [inline utils/config.mk:15]: Similarly to my comment on `Makefile`... what does this do/mean?

**Review by brano2 — COMMENTED** [review]
Looks good, I made some comments and suggestions. I think it would be worthwhile to add more tests as I suggested in one of the comments. Mainly to serve as an example for how to write tests properly. (Currently it might set a bit of a bad example.) But if you don't think it's a good idea, I'll approve once you respond to my comments.

File: Makefile
Comment (mkristien) [reply Makefile:114]: The Makefiles are a bit messy atm, as the QA team is in the midst of development/prototyping for some stuff. It would deserve some refactoring. I think, documenting this does not make sense now, as refactoring will change the structure of Makefiles a lot anyway. I can create a new task in clickup to refactor and document Makefiles. What do you think?

Btw, I think this libtest.a task should go away completely in the refactoring process.

File: test/src/utils/config.test.cpp
Comment (mkristien) [reply test/src/utils/config.test.cpp:68]: Frankly, I do not know how to test these failure cases. The implementation just posts an ERROR message and goes on. Circular include are prevented, with error message indicating any attempts of circular includes. However, the state of the config is the same as if the error line in `config.txt` was not there.

File: Makefile
Comment (brano2) [reply Makefile:114]: Sounds good. Yes, create a ClickUp (sub)task for it please.

Comment (mkristien) [reply Makefile:114]: Task created under `General Improvements and Refactoring` https://app.clickup.com/t/2mnvcz

File: test/src/utils/config.test.cpp
Comment (brano2) [reply test/src/utils/config.test.cpp:68]: Fair enough, I forgot that it was only printing error messages and not actually raising exceptions or anything like that. It's fine then.

File: configurations/config.txt
Comment (mkristien) [reply configurations/config.txt:7]: I agree, good point, will update the comments :)

This, however, illustrates an interesting issue with PR reviews. This code is not really new, i.e. is not a contribution of this PR, as the content of this file was in the repo since the beginning. The PR only moved the file into `configurations` folder. Thus, is it fair to request changes to code that other people wrote before this PR branched off of develop?

Comment (brano2) [reply configurations/config.txt:7]: Good point. But I think in general, when you touch old code (even if you just move it), you should check it and fix any simple problems like this. If it was something more complex then it would be okay to leave any refactoring for a separate PR (and potentially the original author of the code).

Comment (brano2) [reply configurations/config.txt:7]: Also, in this case you extended the comment as well, not just copied it. Of course, if you move a lot of stuff around, it might sometimes not be feasible to review and refactor it all but in this case you made edits right next to the part I commented on. So you probably would have reviewed the existing comment anyway to make sure that what you added actually fits.

**Review by brano2 — APPROVED** [review]

**Pull request merged**
